service/src/test/kotlin/app/cash/backfila/service/BackfillRunnerTest.kt

```diff
@@ -589,6 +588,161 @@ class BackfillRunnerTest {
     }
   }
 
+  @Test fun `returning partial backfill bookmark makes followup runbatch`() {
+    fakeBackfilaClientServiceClient.dontBlockGetNextBatch()
+    val runner = startBackfill(numThreads = 1)
+
+    runBlocking {
+      launch { runner.run() }
+      try {
+        val initialRequest = fakeBackfilaClientServiceClient.runBatchRequests.receive()
+
+        // Managed to complete the first item only
+        val newStart = initialRequest.batch_range.start.utf8().toInt() + 1
+        var remainingRange = initialRequest.batch_range.newBuilder()
+            .start(newStart.toString().encodeUtf8())
+            .build()
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().remaining_batch_range(remainingRange).build()))
+
+        // Follow up request is sent immediately
+        val followup1 = withTimeout(100) {
+          fakeBackfilaClientServiceClient.runBatchRequests.receive()
+        }
+
+        // Batch ranges should be the remaining range
+        assertThat(remainingRange).isEqualTo(followup1.batch_range)
+
+        // Now we complete the last item in the batch
+        val newEnd = followup1.batch_range.end.utf8().toInt() - 1
+        remainingRange = followup1.batch_range.newBuilder()
+            .end(newEnd.toString().encodeUtf8())
+            .build()
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().remaining_batch_range(remainingRange).build()))
+
+        // Follow up request is sent immediately
+        val followup2 = withTimeout(100) {
+          fakeBackfilaClientServiceClient.runBatchRequests.receive()
+        }
+
+        // Batch ranges are still the same but with an updated bookmark
+        assertThat(followup2.batch_range).isEqualTo(remainingRange)
+
+        // Cursor hasn't been updated yet
+        waitForOtherCoroutines()
+        assertThat(getSinglePartitionCursor(runner)).isNull()
+
+        // Batch is done
+        fakeBackfilaClientServiceClient.runBatchResponses.send(
+            Result.success(RunBatchResponse.Builder().build()))
+
+        // Cursor is updated once the batch succeeds
+        waitForOtherCoroutines()
+        assertThat(getSinglePartitionCursor(runner))
+            .isEqualTo(initialRequest.batch_range.end.utf8().toLong())
+
+        val nextBatch = fakeBackfilaClientServiceClient.runBatchRequests.receive()
+        // Batch ranges are no longer the same
+        assertThat(initialRequest.batch_range).isNotEqualTo(nextBatch.batch_range)
+      } finally {
+        runner.stop()
+      }
+    }
+  }
+
+  @Test fun `partial backfill bookmarks reduce parallelism`() {
```

Comment:
maybe add a note about whether this is a desired or non-desired attribute of bookmarks